### PR TITLE
Edar/datasets s3

### PR DESCRIPTION
**Pull request by Efrat1**
## Summary
Support S3 data sources.
Based on https://github.com/securefederatedai/openfl/pull/1434

## Type of Change (Mandatory)
- Feature enhancement  

## Description (Mandatory)
For each S3 data source we save:
URI: path to bucket
endpoint
access_key_env_name: name of environment variable of access key
secret_key_env_name: name of environment variable of secret key
secret_name: for future use
(optional) hash_func: for calculate hash per file. if not set, use S3 Etag instead.

## Testing
Unittests using moto mock_aws
Manual testing with openfl-security using MINIO

### CODE REVIEW

File: openfl/federated/data/sources/s3_data_source.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        uri: str,
+        endpoint=None,
+        aws_access_key_id=None,
+        aws_secret_access_key=None,
+        hash_func=None,
+    ):
+        super().__init__(DataSourceType.S3)
+        self.uri = uri
+        self.endpoint = endpoint
+        self.aws_access_key_id = aws_access_key_id
```
Comment (vrancurel) [inline openfl/federated/data/sources/s3_data_source.py]: access_key_env_name
secret_key_env_name

**Review by vrancurel — CHANGES_REQUESTED** [review]

File: openfl/federated/data/sources/s3_data_source.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        prefix = parsed.path.lstrip("/")  # Remove leading slash
+
+        aws_access_key_id = (
+            os.environ.get(self.aws_access_key_id) if self.aws_access_key_id else None
+        )
+        aws_secret_access_key = (
+            os.environ.get(self.aws_secret_access_key) if self.aws_secret_access_key else None
+        )
+
+        s3_client = boto3.client(
```
Comment (vrancurel) [inline openfl/federated/data/sources/s3_data_source.py]: can be moved to __init__ once and for all

**Review by vrancurel — CHANGES_REQUESTED** [review]

File: openfl/federated/data/sources/s3_data_source.py (hunk 1 above)
Comment (Efrat1) [reply openfl/federated/data/sources/s3_data_source.py]: done

File: openfl/federated/data/sources/s3_data_source.py (hunk 2 above)
Comment (Efrat1) [reply openfl/federated/data/sources/s3_data_source.py]: nice, done

File: openfl/federated/data/sources/s3_data_source.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        uri: str,
+        endpoint=None,
+        access_key_env_name=None,
+        secret_key_env_name=None,
+        hash_func=None,
+    ):
+        super().__init__(DataSourceType.S3)
+        self.uri = uri
+        self.endpoint = endpoint
+        self.access_key_env_name = access_key_env_name
```
Comment (vrancurel) [inline openfl/federated/data/sources/s3_data_source.py:28]: I understand now your config file.
In the S3DataSource there shall be the access_key and secret_key and not the env name.
Can you rename the vars to `access_key` and `secret_key` this is simpler. thx

File: openfl/federated/data/sources/pytorch/__init__.py
Comment (teoparvanov) [inline openfl/federated/data/sources/pytorch/__init__.py:1]: nit - we usually call PyTorch-related packages simply `torch`

File: openfl/federated/data/sources/pytorch/s3_folder.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        """Loads all file paths and their inferred labels into self.samples."""
+        parsed = urlparse(self.uri)
+        bucket_name = parsed.netloc
+        prefix = parsed.path.lstrip("/")
+        access_key_env_name = (
+            os.environ.get(self.access_key_env_name) if self.access_key_env_name else None
+        )
+        secret_key_env_name = (
+            os.environ.get(self.secret_key_env_name) if self.secret_key_env_name else None
+        )
```
Comment (teoparvanov) [inline openfl/federated/data/sources/pytorch/s3_folder.py]: this code section is a bit confusing - are `access_ken_env_name` and `secret_key_env_name` names of env variables, or their content?

File: openfl/federated/data/sources/pytorch/s3_folder.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        endpoint=None,
+        access_key_env_name=None,
+        secret_key_env_name=None,
+        transform=None,
+    ):
+        """
+        Args:
+            uri (str): URI to the S3 object.
+            label_mapper (LabelMapper): LabelMapper object to map class names to indices.
+            transform (callable, optional): Transformations to apply to images.
```
Comment (teoparvanov) [inline openfl/federated/data/sources/pytorch/s3_folder.py]: could you update the doc string to match the actual parameters?

File: openfl/federated/data/sources/pytorch/s3_image_folder.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        parsed = urlparse(obj_path)
+        bucket_name = parsed.netloc
+        object_key = parsed.path.lstrip("/")
+
+        access_key_env_name = (
+            os.environ.get(self.access_key_env_name) if self.access_key_env_name else None
+        )
+        secret_key_env_name = (
+            os.environ.get(self.secret_key_env_name) if self.secret_key_env_name else None
+        )
```
Comment (teoparvanov) [inline openfl/federated/data/sources/pytorch/s3_image_folder.py]: same here - please distinguish between the env name and the value

**Review by teoparvanov — COMMENTED** [review]
Hi @Efrat1 , overall it's looking good, but a bit hard to review in its current state (to some extent because https://github.com/securefederatedai/openfl/pull/1434 isn't merged yet). 

To improve the "reviewer experience, it could be helpful to add an example code snippet of how a (PyTorch) data loader would "plug" into this class hierarchy.

For example, AFAIU, the `torch.utils.data.DataLoader` would be instantiated with a `VerifiableImageFolder` instance, which would in turn iterate through the data using the `__getitem__()` method, right?
image

File: openfl/federated/data/sources/pytorch/__init__.py
Comment (Efrat1) [reply openfl/federated/data/sources/pytorch/__init__.py:1]: will be fixed in https://github.com/securefederatedai/openfl/pull/1434

File: openfl/federated/data/sources/pytorch/s3_image_folder.py (hunk 6 above)
Comment (Efrat1) [reply openfl/federated/data/sources/pytorch/s3_image_folder.py]: done

File: openfl/federated/data/sources/pytorch/s3_folder.py (hunk 5 above)
Comment (Efrat1) [reply openfl/federated/data/sources/pytorch/s3_folder.py]: done

File: openfl/federated/data/sources/pytorch/s3_folder.py (hunk 4 above)
Comment (Efrat1) [reply openfl/federated/data/sources/pytorch/s3_folder.py]: right, done

File: openfl/federated/data/sources/s3_data_source.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        self.hash_func = hash_func
+
+        access_key_id = os.environ.get(access_key_env_name) if access_key_env_name else None
+        secret_access_key = os.environ.get(secret_key_env_name) if secret_key_env_name else None
+
+        self._s3_client = boto3.client(
+            "s3",
+            endpoint_url=self.endpoint,
+            aws_access_key_id=access_key_id,
+            aws_secret_access_key=secret_access_key,
```
Comment (teoparvanov) [inline openfl/federated/data/sources/s3_data_source.py:41]: What would happen here if `access_key_id` or `secret_access_key` are `None` ? Is this the case for publicly available S3 objects?

File: openfl/federated/data/sources/s3_data_source.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
+# Copyright 2020-2025 Intel Corporation
```
Comment (teoparvanov) [inline openfl/federated/data/sources/s3_data_source.py]: This is a new file, therefore the copyright header should only cover 2025 at this point. In case of significant updates to this module (say, in 2026), the header would be changed to `2025-2026`.

File: openfl/federated/data/sources/torch/folder_dataset.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
+# Copyright 2020-2025 Intel Corporation
```
Comment (teoparvanov) [inline openfl/federated/data/sources/torch/folder_dataset.py]: same here: Copyright 2025 Intel Corporation

File: openfl/federated/data/sources/torch/s3_folder.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
+# Copyright 2020-2025 Intel Corporation
```
Comment (teoparvanov) [inline openfl/federated/data/sources/torch/s3_folder.py]: nit: 2025

File: openfl/federated/data/sources/torch/s3_image_folder.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
+# Copyright 2020-2025 Intel Corporation
```
Comment (teoparvanov) [inline openfl/federated/data/sources/torch/s3_image_folder.py]: same here: 2025

File: tests/openfl/federated/data/test_s3_data_source.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
+# Copyright 2020-2025 Intel Corporation
```
Comment (teoparvanov) [inline tests/openfl/federated/data/test_s3_data_source.py]: => simply 2025

File: openfl/federated/data/sources/torch/folder_dataset.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+        """Assigns or retrieves the index of a label."""
+        if label not in self.label_to_idx:
+            new_index = len(self.label_to_idx)
+            self.label_to_idx[label] = new_index
+            self.idx_to_label[new_index] = label
+        return self.label_to_idx[label]
+
+    def get_label_name(self, index: int) -> str:
+        """Retrieves the original label name from an index."""
+        return self.idx_to_label.get(index, None)
```
Comment (teoparvanov) [inline openfl/federated/data/sources/torch/folder_dataset.py]: I see this class is defined in the `sources.torch` package, although there doesn't seem to be anything specific to PyTorch in here. Would it make sense to extract it as a shared utility, that could potentially be reused in other related packages like `sources.tensorflow` too?

File: test-requirements.txt [hunk 14]
```diff
@@ -10,3 +10,5 @@ fpdf2==2.8.2
 papermill==2.6.0
 pyfakefs==5.8.0
 torch==2.4.1
+boto3
+moto
```
Comment (teoparvanov) [inline test-requirements.txt]: please set specific versions here, to avoid breakages when those libraries get new releases

File: openfl/federated/data/sources/torch/s3_folder.py [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+        secret_key = os.environ.get(self.secret_key_env_name) if self.secret_key_env_name else None
+
+        s3_client = boto3.client(
+            "s3",
+            endpoint_url=self.endpoint,
+            aws_access_key_id=access_key,
+            aws_secret_access_key=secret_key,
+        )
+        paginator = s3_client.get_paginator("list_objects_v2")
+        samples = []
```
Comment (teoparvanov) [inline openfl/federated/data/sources/torch/s3_folder.py]: There seems to be some duplication with the logic in `S3DataSource`. Couldn't the `S3Folder` reference a `S3DataSource` instead of directly interacting with S3?

File: openfl/federated/data/sources/torch/s3_image_folder.py [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+import os
+from urllib.parse import urlparse
+
+import boto3
+from PIL import Image
+
+from openfl.federated.data.sources.torch.s3_folder import S3Folder
+
+
+class S3ImageFolder(S3Folder):
```
Comment (teoparvanov) [inline openfl/federated/data/sources/torch/s3_image_folder.py]: How does this class relate to the `VerifiableImageFolder`? Is the `S3ImageFolder` "non-verifiable"?

**Review by teoparvanov — COMMENTED** [review]
Thanks @Efrat1, the PR looks much cleaner after the rebase.
Following up with a couple of additional questions and comments:

File: openfl/federated/data/sources/s3_data_source.py (hunk 8 above)
Comment (Efrat1) [reply openfl/federated/data/sources/s3_data_source.py]: Done

File: openfl/federated/data/sources/torch/folder_dataset.py (hunk 13 above)
Comment (Efrat1) [reply openfl/federated/data/sources/torch/folder_dataset.py]: Good point, anyway, `FolderDataset` does inherit from `torch.utils.data.Dataset`, so it's still PyTorch-specific.
Once we implement the TensorFlow datasets, we can look back at this and see if there’s shared logic worth extracting.

File: test-requirements.txt (hunk 14 above)
Comment (Efrat1) [reply test-requirements.txt]: Thanks, done

File: openfl/federated/data/sources/torch/s3_folder.py (hunk 15 above)
Comment (Efrat1) [reply openfl/federated/data/sources/torch/s3_folder.py]: Good point, I added a small refactor in a separate commit

File: openfl/federated/data/sources/torch/s3_image_folder.py (hunk 16 above)
Comment (Efrat1) [reply openfl/federated/data/sources/torch/s3_image_folder.py]: `VerifiableImageFolder` (a `VerifiableMapStyleDataset`) manages a list of Torch datasets, one per data source. `S3ImageFolder` can be one of those datasets. The actual verification happens inside `VerifiableMapStyleDataset`, not in the individual datasets. So `S3ImageFolder` is simply a Torch dataset that knows how to read images from S3 — it's not responsible for any verification on its own.

File: openfl/federated/data/sources/torch/verifiable_map_style_image_folder.py
Comment (teoparvanov) [inline openfl/federated/data/sources/torch/verifiable_map_style_image_folder.py:1]: @Efrat1, once this PR is finalized, would you have all the building blocks to create an example OpenFL workspace (f.e. `openfl-workspaces/torch/histology_s3`) that loads from `S3DataSource` (without verifying dataset items)?

File: openfl/federated/data/sources/torch/s3_image_folder.py (hunk 16 above)
Comment (teoparvanov) [reply openfl/federated/data/sources/torch/s3_image_folder.py]: I see, thanks for the clarification!

File: openfl/federated/data/sources/data_source.py [hunk 17]
```diff
@@ -48,12 +48,23 @@ def compute_file_hash(self, path: str) -> str:
...
 
     @abstractmethod
-    def enumerate_files(self) -> Generator[str, None, None]:
+    def enumerate_files(self) -> Generator[Tuple[str, str], None, None]:
         """
-        Enumerate all files in the data source.
+        Enumerate all files in the data source and yield their paths and labels.
 
+        Yields:
+            Tuple[str, str]: A tuple containing the file path and its label.
```
Comment (vrancurel) [inline openfl/federated/data/sources/data_source.py]: there is not always a label.

Comment (Efrat1) [reply openfl/federated/data/sources/data_source.py]: fixed

**Review by teoparvanov — APPROVED** [review]
Looks great, thanks @Efrat1 !
